File: Final Artifact/Final Artifact/AnimalShelter/Driver.py
# Import SQLite which was used for the database
import sqlite3

# DateTime libraries imported to allow the new algorithms to retrieve date/time data
from datetime import datetime, timedelta

# Importing the Dog and Monkey classes plus the new database functions that
# replaced the basic array list functions
from Dog import Dog, get_all_dogs, add_dog
from Monkey import Monkey, get_all_monkeys, add_monkey

# create tables for monkeys and dogs in the database if they don't already exist
# this function could be removed, but would need to be run again if the database 
# ever changes location or is deleted. Leaving it in place is harmless since it 
# will not actually run these SQL statements if the tables already exist.
def create_tables():
    conn = sqlite3.connect('rescue_animals.db')
    c = conn.cursor()

    # Create table for dogs
    c.execute('''
        CREATE TABLE IF NOT EXISTS dogs (
            id INTEGER PRIMARY KEY,
            name TEXT,
            breed TEXT,
            gender TEXT,
            age TEXT,
            weight TEXT,
            acquisition_date TEXT,
            acquisition_country TEXT,
            training_status TEXT,
            reserved INTEGER,
            in_service_country TEXT
        )
    ''')

    # Create table for monkeys
    c.execute('''
        CREATE TABLE IF NOT EXISTS monkeys (
            id INTEGER PRIMARY KEY,
            name TEXT,
            species TEXT,
            gender TEXT,
            age TEXT,
            weight TEXT,
            tail_length TEXT,
            height TEXT,
            body_length TEXT,
            acquisition_date TEXT,
            acquisition_country TEXT,
            training_status TEXT,
            reserved INTEGER,
            in_service_country TEXT
        )
    ''')

    # Commit database changes
    conn.commit()
    conn.close()


create_tables()


def main():

    display_menu()


# This method prints the menu options
def display_menu():
    print("\n\n")
    print("\t\t\t\tRescue Animal System Menu")
    print("[1] Intake a new dog")
    print("[2] Intake a new monkey")
    print("[3] Reserve an animal")
    print("[4] Print a list of all dogs")
    print("[5] Print a list of all monkeys")
    print("[6] Print a list of all animals that are not reserved")
    print("[7] Search an animal by name from the database")
    print("[8] Print a list of animals in the database, sorted by age")
    print("[9] Update the training status of all animals based on their tenure")
    print("[0] Delete an animal from the database by their ID number")
    print("[q] Quit application")
    print()
    print("Enter a menu selection")

    user_input = input()
    if user_input == '1':
        intake_new_dog()
    elif user_input == '2':
        intake_new_monkey()
    elif user_input == '3':
        reserve_animal()
    elif user_input == '4':
        print_animals('4')
    elif user_input == '5':
        print_animals('5')
    elif user_input == '6':
        print_animals('6')
    elif user_input == '7':
        print("Enter the name of the animal to search:")
        name = input()
        search_animal_by_name(name)
    elif user_input == '8':
        sort_animals_by_age()
    elif user_input == '9':
        update_training_status()
    elif user_input == '0':
        delete_animal_by_id()
    elif user_input == 'q':
        exit(0)
    else:
        print("Select a valid menu option")  # Input validation for the menu
        display_menu()
        user_input = input()


# Method to add a new dog to the database
def intake_new_dog():
    print("What is the dog's name?")
    name = input()

    dogs = get_all_dogs()
    for dog in dogs:
        if dog[1].lower() == name.lower():
            print("\n\nThis dog is already in our system\n\n")
            return

    print("Breed:")
    breed = input()

    print("Gender:")
    gender = input()

    print("Age:")
    age = input()

    print("Weight:")
    weight = input()

    print("Date acquired:")
    acquisition_date = input()

    print("Country acquired:")
    acquisition_country = input()

    print("Training status:")
    training_status = input()

    print("Reserved? (true/false)")
    reserved = input().lower() == 'true'

    if reserved:
        print("Country of service:")
        in_service_country = input()
    else:
        in_service_country = ""

    dog = Dog(name, breed, gender, age, weight, acquisition_date, acquisition_country, training_status, reserved, in_service_country)
    add_dog(dog)
    print("New dog has been added.")
    display_menu()


# Method to add a new monkey to the database
# This method is similar to the original array list editor, but 
# basic Python commands have been replaced with the proper SQL syntax
def intake_new_monkey():
    print("What is the monkey's name?")
    name = input()

    monkeys = get_all_monkeys()
    for monkey in monkeys:
        if monkey[1].lower() == name.lower():
            print("\n\nThis monkey is already in our system\n\n")
            return

    print("Species: ")
    species = input()
    if species.lower() not in ["capuchin", "guenon", "macaque", "marmoset", "squirrel monkey", "tamarin"]:
        print("We currently are not accepting this species of monkey.")
        return

    print("Gender: ")
    gender = input()

    print("Age: ")
    age = input()

    print("Weight: ")
    weight = input()

    print("Tail length: ")
    tail_length = input()

    print("Height: ")
    height = input()

    print("Body length: ")
    body_length = input()

    print("Acquisition date: ")
    acquisition_date = input()

    print("Acquisition country: ")
    acquisition_country = input()

    print("Training status: ")
    training_status = input()

    print("Reserved? (true/false)")
    reserved = input().lower() == 'true'

    if reserved:
        print("Country of service: ")
        in_service_country = input()
    else:
        in_service_country = ""

    monkey = Monkey(name, species, gender, age, weight, tail_length, height, body_length, acquisition_date, acquisition_country, training_status, reserved, in_service_country)
    add_monkey(monkey)
    print("New monkey has been added.")
    display_menu()


# Method allowing a user to select an animal to reserve from the database
def reserve_animal():
    print("Enter desired animal type:")
    animal_type = input()
    print("Enter desired animal country:")
    in_service_country = input()

    if animal_type.lower() == "dog":
        dogs = get_all_dogs()
        for dog in dogs:
            if dog[10].lower() == in_service_country.lower():

                # Connects to the SQLite DB and sets the cursor so that the queried
                # animal can be found
                conn = sqlite3.connect('rescue_animals.db')
                c = conn.cursor()
                c.execute('UPDATE dogs SET reserved = 1 WHERE id = ?', (dog[0],))
                conn.commit()
                conn.close()
                print(dog[1] + " has been reserved for you.")
                break
        else:
            print("It doesn't look like we have a dog available that matches your criteria.")
            display_menu()

    elif animal_type.lower() == "monkey":
        monkeys = get_all_monkeys()
        for monkey in monkeys:
            if monkey[13].lower() == in_service_country.lower():

                # Connects to the SQLite DB and sets the cursor so that the queried
                # animal can be found
                conn = sqlite3.connect('rescue_animals.db')
                c = conn.cursor()
                c.execute('UPDATE monkeys SET reserved = 1 WHERE id = ?', (monkey[0],))
                conn.commit()
                conn.close()
                print(monkey[1] + " has been reserved for you.")
                break
        else:
            print("It doesn't look like we have a monkey available that matches your criteria.")
            display_menu()
    else:
        print("That is not a valid choice. Grazioso Salvare only trains dogs and monkeys at this time.")
        display_menu()


# Prints a list of animals based on the selection a user makes from the main menu
def print_animals(choice):
    if choice == '4':
        dogs = get_all_dogs()
        for dog in dogs:
            print(f"Name: {dog[1]} Training status: {dog[8]} Reserved: {dog[9]}")
            print("Country: " + dog[10])
            
            # Hyphens used as an attractive separator between lines of results
            print("----------")
        display_menu()
    elif choice == '5':
        monkeys = get_all_monkeys()
        for monkey in monkeys:
            print(f"Name: {monkey[1]} Training status: {monkey[11]} Reserved: {monkey[12]}")
            print("Country: " + monkey[13])
            print("----------")
        display_menu()
    elif choice == '6':
        print("Animals that are not currently reserved:")
        dogs = get_all_dogs()
        for dog in dogs:
            if not dog[9]:
                print(f"Name: {dog[1]} Type: Dog")
                print(f"Training status: {dog[8]} Country: {dog[10]}")
                print("----------")
        monkeys = get_all_monkeys()
        for monkey in monkeys:
            if not monkey[12]:
                print(f"Name: {monkey[1]} Type: Monkey")
                print(f"Training status: {monkey[11]} Country: {monkey[13]}")
                print("----------")
        display_menu()

# Search algorithm that will allow a user to search for a specific animal by name.
# Note that names are case-sensitive.
def search_animal_by_name(name):
    conn = sqlite3.connect('rescue_animals.db')
    c = conn.cursor()

    c.execute("SELECT * FROM dogs WHERE name = ?", (name,))
    dogs = c.fetchall()

    c.execute("SELECT * FROM monkeys WHERE name = ?", (name,))
    monkeys = c.fetchall()

    conn.close()

    if not dogs and not monkeys:
        print(f"No animals found with the name: {name}")
        display_menu()
    else:
        for dog in dogs:
            print(f"Dog found: {dog}")
        for monkey in monkeys:
            print(f"Monkey found: {monkey}")
        display_menu()

# Simple sorting algorithm that will order animals by age. This could 
# be useful to a user or admin that is auditing the lifespan and 
# likelihood of "retirement" of animals in the future.
def sort_animals_by_age():
    conn = sqlite3.connect('rescue_animals.db')
    c = conn.cursor()

    c.execute("SELECT * FROM dogs ORDER BY age")
    dogs = c.fetchall()

    c.execute("SELECT * FROM monkeys ORDER BY age")
    monkeys = c.fetchall()

    conn.close()

    print("Dogs sorted by age:")
    for dog in dogs:
        print(dog)

    print("-------------------------")
    print("Monkeys sorted by age:")
    for monkey in monkeys:
        print(monkey)
    
    display_menu()

# More involved algorithm that iterates through the database and automatically assigns
# a training status value based on how long the animal has been in training or in
# the database of available animals. This is helpful because an admin can enter as many
# animals as they want at a time, then the program will automatically sort them based
# on how long they have been in the system and assign them training statuses accordingly.

def update_training_status():
    conn = sqlite3.connect('rescue_animals.db')
    c = conn.cursor()

    # Define the training duration for each phase
    training_phases = [
        ("intake", timedelta(days=1)),
        ("Phase I", timedelta(days=30)),
        ("Phase II", timedelta(days=60)),
        ("in service", timedelta(days=120))
    ]

    # Ensure phases are sorted by duration
    training_phases.sort(key=lambda x: x[1])

    # Internal definition for the status updates
    # Several print statements that I used for debugging have been commented out,
    # otherwise there would be hundreds of lines of console output every time this 
    # menu option is called/selected.
    def update_animal_training_status(animal_type, animals):
        for animal in animals:
            animal_id, acquisition_date_str, current_phase = animal
            # print(f"{animal_type.capitalize()} ID: {animal_id}, Acquisition Date: {acquisition_date_str}, Current Phase: {current_phase}")

            try:
                acquisition_date = datetime.strptime(acquisition_date_str, '%m-%d-%Y')
            except ValueError:
                # print(f"Invalid date format for {animal_type} with ID {animal_id}: {acquisition_date_str}")
                continue

            time_in_training = datetime.now() - acquisition_date
            # print(f"Time in Training for {animal_type.capitalize()} ID {animal_id}: {time_in_training.days} days")

            new_phase = current_phase

            for phase, duration in training_phases:
                if time_in_training >= duration:
                    new_phase = phase
                    # print(f"{animal_type.capitalize()} ID {animal_id} eligible for phase: {phase}")

            if new_phase != current_phase:
                # print(f"Updating {animal_type.capitalize()} ID {animal_id} from {current_phase} to {new_phase}")
                c.execute(f"UPDATE {animal_type}s SET training_status = ? WHERE id = ?", (new_phase, animal_id))

    # Update training status for dogs
    c.execute("SELECT id, acquisition_date, training_status FROM dogs")
    dogs = c.fetchall()
    update_animal_training_status('dog', dogs)

    # Update training status for monkeys
    c.execute("SELECT id, acquisition_date, training_status FROM monkeys")
    monkeys = c.fetchall()
    update_animal_training_status('monkey', monkeys)

    conn.commit()
    conn.close()

    print("Training statuses have been updated.")
    display_menu()

# A basic algorithm that allows a user or admin to search an animal 
# and delete them by their unique ID number.
def delete_animal_by_id():
    print("Enter the type of animal to delete (dog/monkey):")
    animal_type = input().lower()
    print("Enter the id of the animal to delete:")
    animal_id = input()

    conn = sqlite3.connect('rescue_animals.db')
    c = conn.cursor()

    if animal_type == 'dog':
        c.execute("DELETE FROM dogs WHERE id = ?", (animal_id,))
    elif animal_type == 'monkey':
        c.execute("DELETE FROM monkeys WHERE id = ?", (animal_id,))
    else:
        print("Invalid animal type.")
        conn.close()
        return

    conn.commit()
    conn.close()
    print(f"{animal_type.capitalize()} with id {animal_id} has been deleted.")
    display_menu()

# Run the program
if __name__ == "__main__":
    main()
    display_menu()

File: Original Artifact/Grazioso/Grazioso/src/Driver.java
import java.util.ArrayList;
import java.util.Scanner;

public class Driver {

    // Creating the array lists for Dog and Monkey
    private static ArrayList<Dog> dogList = new ArrayList<Dog>();
    private static ArrayList<Monkey> monkeyList = new ArrayList<Monkey>();

    public static void main(String[] args) {

        initializeDogList();
        initializeMonkeyList();
        displayMenu();

        // Setting up the main menu options to reference the other methods

        Scanner scnr = new Scanner(System.in);
        char input = scnr.nextLine().charAt(0);
        switch (input) {
            case '1' -> intakeNewDog(scnr);
            case '2' -> intakeNewMonkey(scnr);
            case '3' -> reserveAnimal(scnr);
            case '4' -> printAnimals('4');
            case '5' -> printAnimals('5');
            case '6' -> printAnimals('6');
            case 'q' -> System.exit(0);
            default -> {
                System.out.println("Select a valid menu option"); // Input validation for the menu
                displayMenu();
            }
        }
    }

    // This method prints the menu options
    public static void displayMenu() {
        System.out.println("\n\n");
        System.out.println("\t\t\t\tRescue Animal System Menu");
        System.out.println("[1] Intake a new dog");
        System.out.println("[2] Intake a new monkey");
        System.out.println("[3] Reserve an animal");
        System.out.println("[4] Print a list of all dogs");
        System.out.println("[5] Print a list of all monkeys");
        System.out.println("[6] Print a list of all animals that are not reserved");
        System.out.println("[q] Quit application");
        System.out.println();
        System.out.println("Enter a menu selection");
    }


    // Adds dogs to a list for testing
    public static void initializeDogList() {
        Dog dog1 = new Dog("Spot", "German Shepherd", "male", "1", "25.6", "05-12-2019", "United States", "intake", false, "United States");
        Dog dog2 = new Dog("Rex", "Great Dane", "male", "3", "35.2", "02-03-2020", "United States", "Phase I", false, "United States");
        Dog dog3 = new Dog("Bella", "Chihuahua", "female", "4", "25.6", "12-12-2019", "Canada", "in service", true, "Canada");

        dogList.add(dog1);
        dogList.add(dog2);
        dogList.add(dog3);
    }


    // Adds monkeys to a list for testing
    public static void initializeMonkeyList() {
        Monkey monkey1 = new Monkey("Jack", "Marmoset", "male", "3", "12", "14", "34", "40", "03-21-2018", "Peru", "in service", false, "United States");

        monkeyList.add(monkey1);
    }


    // Complete the intakeNewDog method
    // The input validation to check that the dog is not already in the list
    // is done for you
    public static void intakeNewDog(Scanner scnr) {

        System.out.println("What is the dog's name?");
        String name = scnr.nextLine();
        for (Dog dog : dogList) {
            if (dog.getName().equalsIgnoreCase(name))
            {
                System.out.println("\n\nThis dog is already in our system\n\n");
                return; //returns to menu
            }
            else {

                // Referencing the mutators in class RescueAnimal

                dog.setName(name);

                System.out.println("Breed:");
                dog.setBreed(scnr.nextLine());

                System.out.println("Gender:");
                dog.setGender(scnr.nextLine());

                System.out.println("Age:");
                dog.setAge(scnr.nextLine());

                System.out.println("Weight:");
                dog.setWeight(scnr.nextLine());

                System.out.println("Country acquired:");
                dog.setAcquisitionDate(scnr.nextLine());

                System.out.println("Date acquired:");
                dog.setAcquisitionLocation(scnr.nextLine());

                System.out.println("Training status:");
                dog.setTrainingStatus(scnr.nextLine());

                System.out.println("Reserved? (true/false)");
                dog.setReserved(scnr.nextBoolean());
                scnr.nextLine();

                if (dog.getReserved()) {
                    System.out.println("Country of service:");
                    dog.setInServiceCountry(scnr.nextLine());
                }
                System.out.println("New dog has been added.");

                displayMenu();
                return;
            }
        }
    }
    // Add the code to instantiate a new dog and add it to the appropriate list



        // Complete intakeNewMonkey
	//Instantiate and add the new monkey to the appropriate list
        // For the project submission you must also  validate the input
	// to make sure the monkey doesn't already exist and the species type is allowed
    public static void intakeNewMonkey(Scanner scnr) {

        System.out.println("What is the monkey's name?");
        String name = scnr.nextLine();
        for(Monkey monkey: monkeyList) {
            if(monkey.getName().equalsIgnoreCase(name)) {
            	System.out.println("\n\nThis monkey is already in our system\n\n");
            	return;
            }
            else
            {
                // Check the monkey species to make sure it is acceptable to Grazioso
                System.out.println("Species: ");
                monkey.setSpecies(scnr.nextLine());
                if(!(monkey.getSpecies().equalsIgnoreCase("Capuchin")) && !(monkey.getSpecies().equalsIgnoreCase("Guenon")) && !(monkey.getSpecies().equalsIgnoreCase("Macaque")) && !(monkey.getSpecies().equalsIgnoreCase("Marmoset")) && !(monkey.getSpecies().equalsIgnoreCase("Squirrel Monkey")) && !(monkey.getSpecies().equalsIgnoreCase("Tamarin")))
                {
                    System.out.println("We currently are not accepting this species of monkey.");
                    return;
                }

                // Reference mutators from class RescueAnimal to set attributes

                monkey.setName(name);

                System.out.println("Gender: ");
                monkey.setGender(scnr.nextLine());

                System.out.println("Age: ");
                monkey.setAge(scnr.nextLine());

                System.out.println("Weight: ");
                monkey.setWeight(scnr.nextLine());

                System.out.println("Tail length: ");
                monkey.setTailLength(scnr.nextLine());

                System.out.println("Height: ");
                monkey.setHeight(scnr.nextLine());

                System.out.println("Body length: ");
                monkey.setBodyLength(scnr.nextLine());

                System.out.println("Acquisition date: ");
                monkey.setAcquisitionDate(scnr.nextLine());

                System.out.println("Acquisition country: ");
                monkey.setAcquisitionLocation(scnr.nextLine());

                System.out.println("Training status: ");
                monkey.setTrainingStatus(scnr.nextLine());

                System.out.println("Reserved status: ");
                boolean reserved = scnr.nextBoolean();

                if (monkey.getReserved()) {
                    System.out.println("Country of service: ");
                    monkey.setInServiceCountry(scnr.nextLine());
                }
                System.out.println("New monkey has been added.");
                displayMenu();
                return;
            }
        }
    }

        // Complete reserveAnimal
        // You will need to find the animal by animal type and in service country
        public static void reserveAnimal(Scanner scnr) {
            System.out.println("Enter desired animal type: ");       // Querying user for animal type
            String type = scnr.nextLine();
            System.out.println("Enter desired animal country: ");    // Querying user for animal country
            String inServiceCountry = scnr.nextLine();
            if(type.equalsIgnoreCase("Dog"))
            {
                for (Dog dog : dogList) {
                    if (dog.getInServiceCountry().equals(inServiceCountry)) {
                        // Updating the reserved status
                        dog.setReserved(true);
                        System.out.println(dog.getName() + " has been reserved for you.");
                        break;
                    }
                }
                System.out.println("It doesn't look like we have a dog available that matches your criteria.");
                System.out.println("Please check your spelling or try another country.");
                displayMenu();
            }
            else if(type.equalsIgnoreCase("Monkey"))
            {
                for(Monkey monkey : monkeyList) {
                    if (monkey.getInServiceCountry().equals(inServiceCountry)) {
                        // Updating the reserved status
                        monkey.setReserved(true);
                        System.out.println(monkey.getName() + " has been reserved for you.");
                        break;
                    }
                    else {
                        System.out.println("It doesn't look like we have a monkey available that matches your criteria.");
                        System.out.println("Please check your spelling or try another country.");
                        displayMenu();
                    }
                }
            }
            else {
                System.out.println("That is not a valid choice. Grazioso Salvare only trains dogs and monkeys at this time.");
                System.out.println("Please check your spelling or enter a valid choice (dog or monkey).");
                displayMenu();
            }

        }

        // Complete printAnimals
        // Include the animal name, status, acquisition country and if the animal is reserved.
	// Remember that this method connects to three different menu items.
        // The printAnimals() method has three different outputs
        // based on the listType parameter
        // dog - prints the list of dogs
        // monkey - prints the list of monkeys
        // available - prints a combined list of all animals that are
        // fully trained ("in service") but not reserved 
	// Remember that you only have to fully implement ONE of these lists. 
	// The other lists can have a print statement saying "This option needs to be implemented".
	// To score "exemplary" you must correctly implement the "available" list.
        public static void printAnimals(char choice)
        {
            if(choice == '4')
            {                                         // Displays name, training status, reserved status, and country
                for (Dog dog : dogList) {
                    System.out.printf("%-5s %-10s %-7s %-10s %-9s %-10s %n", "Name:", dog.getName(), "Training status:", dog.getTrainingStatus(), "Reserved:", dog.getReserved());
                    System.out.println("Country: " + dog.getInServiceCountry());
                    System.out.println("----------");
                }
                displayMenu();
            }
            else if(choice == '5')
            {                                        // Displays name, training status, reserved status, and country
                for (Monkey monkey : monkeyList) {
                    System.out.printf("%-5s %-10s %-7s %-10s %-9s %-10s %n", "Name:", monkey.getName(), "Training status:", monkey.getTrainingStatus(), "Reserved:", monkey.getReserved());
                    System.out.println("Country: " + monkey.getInServiceCountry());
                    System.out.println("----------");
                }
                displayMenu();
            }
            else if(choice == '6')
            {
                System.out.println("Animals that are not currently reserved:");
                for (Dog dog : dogList) {
                    if(!dog.getInServiceCountry().isEmpty() && !dog.getReserved())
                    {
                        System.out.printf("%-5s %-10s %-5s %n", "Name:", dog.getName(), "Type: Dog");
                        System.out.println("Training status: " + dog.getTrainingStatus() + "   Country: " + dog.getInServiceCountry());
                        System.out.println("----------");
                    }
                }
                for (Monkey monkey : monkeyList)
                {
                    if(!monkey.getInServiceCountry().isEmpty() && !monkey.getReserved())
                    {
                        System.out.printf("%-5s %-10s %-5s %n", "Name:", monkey.getName(), "Type: Monkey");
                        System.out.println("Training status: " + monkey.getTrainingStatus() + "   Country: " + monkey.getInServiceCountry());
                        System.out.println("----------");
                    }
                }
                displayMenu();
            }
        }
}

